### PR TITLE
[SPARK-41631][FOLLOWUP][SQL] Fix two issues in implicit lateral column alias resolution on Aggregate

### PR DESCRIPTION
**Pull request by anchovYu**
### What changes were proposed in this pull request?

Fix two issues in implicit lateral column alias resolution of Aggregate; added related test cases.
* The check condition whether to lift up in Aggregate is incorrect. Leaf expressions, e.g. `now()` will fail the check and won't be lift up. Changed to follow the code pattern in checkAnalysis.
* Another condition that requires the new Aggregate expressions to be non-empty to lift up in Aggregate is unnecessary. Think about the Aggregate on expressions without any grouping expressions or aggregate expressions, e.g. `select 1 as a, a + 1 .. group by ..`.

### Why are the changes needed?

Fix the bug mentioned above.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

New tests.

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveLateralColumnAliasReference.scala [hunk 1]
```diff
@@ -210,14 +209,10 @@ object ResolveLateralColumnAliasReference extends Rule[LogicalPlan] {
...
           }
-          if (newAggExprs.isEmpty) {
-            agg
-          } else {
-            Project(
-              projectList = projectExprs,
-              child = agg.copy(aggregateExpressions = newAggExprs.toSeq)
-            )
-          }
+          Project(
```
Comment (anchovYu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveLateralColumnAliasReference.scala:212]: One thing I'm not sure about is if the code path for empty Aggregate expressions e.g. `Aggregate [dept] []` exists.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveLateralColumnAliasReference.scala [hunk 2]
```diff
@@ -168,19 +168,18 @@ object ResolveLateralColumnAliasReference extends Rule[LogicalPlan] {
             && aggregateExpressions.exists(_.containsPattern(LATERAL_COLUMN_ALIAS_REFERENCE)) =>
 
           // Check if current Aggregate is eligible to lift up with Project: the aggregate
-          // expression only contains: 1) aggregate functions, 2) grouping expressions, 3) lateral
-          // column alias reference or 4) literals.
+          // expression only contains: 1) aggregate functions, 2) grouping expressions, 3) leaf
+          // expressions excluding attributes not in grouping expressions
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveLateralColumnAliasReference.scala:172]: what's the actual change? do we allow something new?

Comment (anchovYu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveLateralColumnAliasReference.scala:172]: Check the tests: `Leaf expression as aggregate expressions should be eligible to lift up`.

Comment (anchovYu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveLateralColumnAliasReference.scala:172]: tldr, previously it doesn't allow leaf expressions other than Literal and LateralColumnAliasReference, now it allows.

File: sql/core/src/test/scala/org/apache/spark/sql/LateralColumnAliasSuite.scala [hunk 3]
```diff
@@ -547,7 +547,8 @@ class LateralColumnAliasSuite extends LateralColumnAliasSuiteBase {
 
   test("Lateral alias of a complex type") {
     // test both Project and Aggregate
-    val querySuffixes = Seq("", s"FROM $testTable GROUP BY dept HAVING dept = 6")
+    // TODO(anchovyu): re-enable aggregate tests when fixed the having issue
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/LateralColumnAliasSuite.scala:550]: but it worked before this PR?

Comment (anchovYu) [reply sql/core/src/test/scala/org/apache/spark/sql/LateralColumnAliasSuite.scala:550]: It was already broken (this is the issue that I would like to discuss with you last week), and it just happened to work due to the twist in the resolution order.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveLateralColumnAliasReference.scala [hunk 4]
```diff
@@ -168,19 +168,18 @@ object ResolveLateralColumnAliasReference extends Rule[LogicalPlan] {
...
           // This check is to prevent unnecessary transformation on invalid plan, to guarantee it
           // throws the same exception. For example, cases like non-aggregate expressions not
           // in group by, once transformed, will throw a different exception: missing input.
           def eligibleToLiftUp(exp: Expression): Boolean = {
             exp match {
               case e if AggregateExpression.isAggregate(e) => true
-              case e if groupingExpressions.exists(_.semanticEquals(e)) => true
-              case _: Literal | _: LateralColumnAliasReference => true
+              case e: Attribute if !groupingExpressions.exists(_.semanticEquals(e)) => false
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveLateralColumnAliasReference.scala:179]: can we move it closer to `case e if groupingExpressions.exists(_.semanticEquals(e)) => true`?

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveLateralColumnAliasReference.scala:179]: and do we really need both cases? I think
```
case e if groupingExpressions.exists(_.semanticEquals(e)) => true
case a: Attribute => false
```
should be sufficient?

Comment (anchovYu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveLateralColumnAliasReference.scala:179]: I follow (copy) the condition in checkAnalysis, but yes I think yours is sufficient. Will change.

**Comment by cloud-fan** [comment]
The failure is unrelated: `python/pyspark/sql/connect/client.py:25: error: Skipping analyzing "grpc_status": module is installed, but missing library stubs or py.typed marker`

**Review by cloud-fan — APPROVED** [review]

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Pull request closed without merging**